lib/Basics/RecursiveLocker.h

```diff
@@ -36,70 +37,81 @@ namespace arangodb {
 template <typename T>
 class RecursiveMutexLocker {
  public:
-  RecursiveMutexLocker( // recursive locker
-      T& mutex, // mutex
-      std::atomic<std::thread::id>& owner, // owner
-      arangodb::basics::LockerType type, // locker type
-      bool acquire, // acquire flag
-      char const* file, // file
-      int line // line
-): _locker(&mutex, type, false, file, line), _owner(owner), _update(noop) {
+  RecursiveMutexLocker(T& mutex, 
+                       std::atomic<std::thread::id>& owner, 
+                       arangodb::basics::LockerType type, 
+                       bool acquire, 
+                       char const* file, 
+                       int line)
+      : _locked(false),
+        _locker(&mutex, type, false, file, line), // does not lock yet 
+        _owner(owner), 
+        _update(noop) {
     if (acquire) {
       lock();
     }
   }
 
   ~RecursiveMutexLocker() { unlock(); }
 
-  bool isLocked() { return _locker.isLocked(); }
+  bool isLocked() const noexcept { return _locked; }
 
   void lock() {
     // recursive locking of the same instance is not yet supported (create a new instance instead)
     TRI_ASSERT(_update != owned);
 
     if (std::this_thread::get_id() != _owner.load()) {  // not recursive
       _locker.lock();
-      _owner.store(std::this_thread::get_id());
-      _update = owned;
+      // it is possible that we cannot acquire the mutex, e.g.
+      // if we are a TRY locker
+      if (_locker.isLocked()) {
+        _owner.store(std::this_thread::get_id());
+        _update = owned;
+        _locked = true;
+      }
+    } else {
+      // we are the owning thread
+      _locked = true;
```

Comment:
`_locker.lock()` only returns once we have acquired the lock. The `LockerType` is only relevant for the constructor.
```suggestion
      _owner.store(std::this_thread::get_id());
      _update = owned;
```